<a href="https://colab.research.google.com/github/KevinAkat/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Kevin_Akat_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [0]:
### YOUR CODE STARTS HERE

from google.colab import files
uploaded = files.upload()

Saving house-votes-84.data to house-votes-84.data


In [30]:
import pandas as pd
import numpy as np

column_headers = ['Party','handicapped-infants','water-project',
                  'adopting-budget-res','physician-fee-freeze',
                  'el-salvador-aid','religious-groups-in-schools',
                  'anti-satellite-test-ban','aid-to-nicaraguan-contras',
                  'mix-missile','immigration','synfuels-corporation cutback',
                  'education-spending','superfund-right-to-sue','crime',
                  'duty-free-exports','south-africa']

df_votes = pd.read_csv('house-votes-84.data', names=column_headers)

print(df_votes.shape)

df_votes.head(10)

(435, 17)


Party handicapped-infants  ... duty-free-exports south-africa
0  republican                   n  ...                 n            y
1  republican                   n  ...                 n            ?
2    democrat                   ?  ...                 n            n
3    democrat                   n  ...                 n            y
4    democrat                   y  ...                 y            y
5    democrat                   n  ...                 y            y
6    democrat                   n  ...                 y            y
7  republican                   n  ...                 ?            y
8  republican                   n  ...                 n            y
9    democrat                   y  ...                 ?            ?

[10 rows x 17 columns]

In [0]:
df_votes = df_votes.replace({'n': 0,'y': 1,'?': np.NaN})

df_votes.head()

Party  handicapped-infants  ...  duty-free-exports  south-africa
0  republican                  0.0  ...                0.0           1.0
1  republican                  0.0  ...                0.0           NaN
2    democrat                  NaN  ...                0.0           0.0
3    democrat                  0.0  ...                0.0           1.0
4    democrat                  1.0  ...                1.0           1.0

[5 rows x 17 columns]

###2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01###

**Null:** Democrats do not support adopting the budget resolution more than republicans
(Dem x-bar == Rep x-bar for adopting budget resolution column)

**Alt:**  Democrate do not support adopting the budget resolution more than republicans (Dem x-bar != Rep x-bar for adopting resolution column)

**Confidence Interval** = 99%
**Alpha** = .01

In [0]:
from scipy import stats

# Seperating dems and reps into 2 differnet DFs

rep = df_votes[df_votes['Party']=='republican']

dem = df_votes[df_votes['Party']=='democrat']

rep['adopting-budget-res'].value_counts()


stats.ttest_ind(dem['adopting-budget-res'], rep['adopting-budget-res'], nan_policy='omit')


Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

***Since p (~0) < alpha (.05), I reject the null hypothesis. In other words, I reject the notion that Democrats do not support the budget resolution more than Republicans.***

###3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01###

**Null:** Republicans do not support The Education Spending Bill more than Democrats (rep x-bar == dem x-bar for educaiton spending column)

**Alt:** Resplublican support The Education Spending Bill more than Demoncrats
(rep x-bar == dem x-bar for education spending column)

**Confidence Interval:** 99%
**Alpha: ** .01

In [56]:

print(rep['adopting-budget-res'].mean())
print(dem['adopting-budget-res'].mean())

stats.ttest_ind(rep['education-spending'], dem['education-spending'], nan_policy='omit')

0.13414634146341464
0.8884615384615384


Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

***Since p (~0) < alpha (.10), I reject the null hypothesis. In other words, I reject the notion that Republicans do not support The Education Bill more than Democrats.***

###4. find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)###

**Null:** Republicans and Democrats have voted equally on The Water Bill
(Republican x-bar == Democrat x-bar for The Water Bill column)

**Alt:** Republicans and Democrate did not vote equally on The Water Bill
(Republican x-bar != Democrate x-bar for The Water Bill column)

**Confidence Interval** = 90%
 **Alpha** = .10

In [58]:
print(rep['water-project'].mean())
dem['water-project'].mean()

stats.ttest_ind(rep['water-project'],dem['water-project'],nan_policy='omit')

0.5067567567567568


Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

**Since p (.929) > alpha (.05), I fail to reject the Null Hypothesis. In other words, I fail to reject the notion that the Democrats and Republicans are equally split on The Water Bill.**

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!

###Stretch Goal 2 - 1-sample t-*test*### 

In [134]:
weighted_dice_probability_of_rolling_1 = 1/3 # weighted dice probability of rolling a 1 (stinkin' cheaters)
fair_dice_probability_of_rolling_1 = 1/6 # population parameter

roll_dice = np.random.binomial(n=1,p=weighted_dice_probability_of_rolling_1,size=25)

mean_dice_sample = roll_dice.mean()
mean_dice_sample

0.4

**Null (Ho)** = The sample die is a fair die. In other words, the die being used for sample testing has the same probability of rolling a 1 as any other fair die.

**Alternative (Ha)** The sample die is not a fair dic. In other words, the die being used for sample testing has the same probability of rolling a 1 as any other fair die.

**Confidence Interval:**  = 95%
**Alpha** = .05

In [142]:
# 1) Here's all the metrics we'll need:

std_roll_dice_sample = np.std(roll_dice) # standard deviation of the sample (s)
n = 25 # sample size used above (n)
mean_dice_sample # This is equal to 0.32 as expressed in the code cells above. This is the mean of the sample (x-bar)
population_mean = fair_dice_probability_of_rolling_1 # This is the population paramater for the mean (mu)

# 2) Calculate the Standard Error

# Stadard error formula = s/sqrt(n)

standard_error = std_roll_dice_sample/np.sqrt(n)


# 3) Calculate the t-score:

# T-score for 1 sample test = (x-bar - mu)/standard error

t_score = (mean_dice_sample - population_mean)/standard_error
t_score


2.3814483610392014

In [153]:
stats.ttest_1samp(roll_dice, population_mean)


Ttest_1sampResult(statistic=2.3333333333333335, pvalue=0.028337217532832647)

In [139]:
sq_root = np.sqrt(n)
sq_root

5.0

In [137]:
std_roll_dice_sample

0.4898979485566356

In [148]:
std_error = std_roll_dice_sample/sq_root
std_error

0.09797958971132711

In [151]:
difference = mean_dice_sample - population_mean
difference

0.23333333333333336

In [152]:
difference/std_error

2.3814483610392014

**Since p (.02833) < alpha (.05), I reject The Null Hypothesis. In other words, I reject the notion that the sample die used for this experiment is a fair a die with an equal probability of rolling a 1 when compared to any other fair die.**

###Stretch Goal 2 - 2-sample t-*test*### 

# Class 1 has 